### PR TITLE
Adjust loot generation rate

### PR DESCRIPTION
**Pull request by mugling**
@Mshock777 continues to make good progress on converting the JSON mapgen from `place_items` to `place_loot`. The latter has much more predictable characteristics.

This PR extends `place_loot` to better handle `ITEM_SPAWNRATE`. Hopefully combined with the migration to the new format along with the new item groups being defined under `itemgroups/` we can finally have both realistic and player-configurable loot generation.

The semantics for an `ITEM_SPAWNRATE` of `2.0` are shown below:

| base chance | base repeat | nett chance | nett repeat |
| --- | --- | --- | --- |
| 100 | 1 | 100 | 1 |
| 30 | 1 | 60 | 1 |
| 80 | 1 | 100 | 1 |
| 100 | [2,4] | 100 | [4,8] |
| 30 | [2,4] | 30 | [4,8] |
| 80 | [1,2] | 80 | [2,4] |

And for an `ITEM_SPAWNRATE` of `0.5`:

| base chance | base repeat | nett chance | nett repeat |
| --- | --- | --- | --- |
| 100 | 1 | 50 | 1 |
| 30 | 1 | 15 | 1 |
| 80 | 1 | 40 | 1 |
| 100 | [2,4] | 100 | [1,2] |
| 30 | [2,4] | 30 | [1,2] |
| 80 | [1,2] | 80 | 1 |

The algorithm prefers to scale `repeat` for entries which specify it. If `repeat` is unspecified (defaults to 1) then `chance` is used **except** where `chance == 100` as this is used extensively throughout the mapgen to guarantee placement of an item.

Presuming there are no objections to this I'm going to start writing documentation for `place_loot` and mark `place_items` as deprecated. There is of course the thorny issue as to whether I should provide a version for builtin (C++) mapgens. I'm somewhat reluctant as to the latter as it discourages their migration and in any case that would be a separate PR.

In a further PR I'm going to provide some hooks for mods to adjust the relative frequencies of different `itemgroup` which should allow us to produce mods like 'More frequent ammo' with ease

### CODE REVIEW

File: src/mapgen.cpp [hunk 1]
```diff
@@ -835,33 +835,38 @@ class jmapgen_item_group : public jmapgen_piece {
...
                 jsi.throw_error( "no such item group", "group" );
             }
             if( !name.empty() && !item_controller->has_template( name ) ) {
                 jsi.throw_error( "no such item", "item" );
             }
+            if( ammo < 0 || ammo > 100 ) {
+                jsi.throw_error( "ammo chance out of range", "ammo" );
+            }
+            if( magazine < 0 || magazine > 100 ) {
+                jsi.throw_error( "ammo chance out of range", "ammo" );
```
Comment (illi-kun) [inline src/mapgen.cpp]: This message is absolutely the same as above, it's probably should mention "magazine" instead of "ammo".

Comment (mugling) [inline src/mapgen.cpp]: Copy-pasta error

**Comment by Mshock777** [comment]
Nice job.
1. What if there is a zero in "repeat"?
2. What if I want to have only one item generally, and more with higher spawn rate?
3. In the last example of 0.5 spawn rate, average amount of items should be able to go below 1. (maybe decrease nett chance if this happens?)
4. Wouldn't it be better to just multiply the highest (or lowest if spawn scale is below 1) by double the spawn scale plus the lowest number, so we get the same average amount of items, but have a bigger range?

**Comment by mugling** [comment]
Can you give examples (eg. with numbers) for parts 3 and 4

**Comment by Mshock777** [comment]
3. If we have **`"repeat": [1,2]`**, average amount of items is **1.5**. With **0.5** spawn scale we will get avg. amount of **1**, but should get **0.75**. Since we don't want to work with decimals, when avg. amount goes below **1**, we can simply multiply base chance by this average amount. (if we have **0.75** average amount of items with **80%** spawn chance, we cant urn it into **1** item with **60%** spawn chance)

**Comment by Mshock777** [comment]
4: If we have `"repeat": [1,2]` we can turn it in `"repeat": [1,5]` instead of `"repeat": [2,4]`.
Average amount is the same, but we get better random range imo

**Comment by mugling** [comment]
> Average amount is the same, but we get better random range imo

The whole purpose of `ITEM_SPAWNRATE` is to skew the numbers upwards/downward, not increase the possible probability range for item spawns.

**Comment by Rivet-the-Zombie** [comment]
@mugling, it's unrelated, but I have to give you a thumbs up for proper usage of 'nett'.

**Pull request merged**